### PR TITLE
Update to ModelingToolkit v4.5

### LINKED ISSUE #2411: Update to MT v4
- no `Operation`: `typeof(@variables x)` is `Tuple{Num}`

### PR DESCRIPTION
**Pull request by mforets**
Closes #2411.

### CODE REVIEW

File: src/Initialization/init_ModelingToolkit.jl [hunk 1]
```diff
@@ -21,22 +22,24 @@ eval(quote
...
+    function _vec(vars::NTuple{L, Union{<:Num, <:Vector{Num}}}) where {L}
         return collect(reduce(vcat, vars))
     end
 
     # case with a single variable
-    _vec(vars::Tuple{Operation}) = [vars[1]]
+    _vec(vars::Tuple{Num}) = [vars[1]]
 
+    _get_variables(expr::Term) = ModelingToolkit.get_variables(expr)
+    _get_variables(expr::Vector{<:Term}) = unique(reduce(vcat, get_variables(ex) for ex in expr))
```
Comment (mforets) [inline src/Initialization/init_ModelingToolkit.jl:42]:
```suggestion
    _get_variables(expr::Vector{<:Term}) = unique(reduce(vcat, _get_variables(ex) for ex in expr))
```

File: src/Initialization/init_ModelingToolkit.jl [hunk 2]
```diff
@@ -21,22 +22,24 @@ eval(quote
...
+    function _vec(vars::NTuple{L, Union{<:Num, <:Vector{Num}}}) where {L}
         return collect(reduce(vcat, vars))
     end
 
     # case with a single variable
-    _vec(vars::Tuple{Operation}) = [vars[1]]
+    _vec(vars::Tuple{Num}) = [vars[1]]
 
+    _get_variables(expr::Term) = ModelingToolkit.get_variables(expr)
+    _get_variables(expr::Vector{<:Term}) = unique(reduce(vcat, _get_variables(ex) for ex in expr))
```
Comment (mforets) [inline src/Initialization/init_ModelingToolkit.jl:42]: Tip (thanks to @isaacsas in [this comment](https://github.com/SciML/ModelingToolkit.jl/issues/731#issuecomment-758990856)): use the in-place version `ModelingToolkit.get_variables!` to accumulate the unique variables on the same array.

File: src/Initialization/init_ModelingToolkit.jl [hunk 3]
```diff
@@ -1,17 +1,18 @@
...
    """
-       _vec(vars::NTuple{L, Union{<:Operation, <:Vector{Operation}}}) where {L}
+       _vec(vars::NTuple{L, Union{<:Num, <:Vector{Num}}}) where {L}
 
    Transform a tuple of operations into one vector of operations.
 
    ### Input
 
-   - `vars` -- tuple where each element is either an `Operation` or a vector of operations
+   - `vars` -- tuple where each element is either variable like (`Num`) or a
```
Comment (schillic) [inline src/Initialization/init_ModelingToolkit.jl:14]:
```suggestion
   - `vars` -- tuple where each element is either a variable-like (`Num`) or a
```
Is that what you mean?

File: src/Initialization/init_ModelingToolkit.jl (hunk 2 above)
Comment (schillic) [reply src/Initialization/init_ModelingToolkit.jl:42]: So do you want to change it?

File: src/Sets/HalfSpace.jl [hunk 4]
```diff
@@ -529,15 +529,15 @@ return quote
 # returns `(true, sexpr)` if expr represents a half-space,
 # where sexpr is the simplified expression sexpr := LHS - RHS <= 0
 # otherwise, returns `(false, expr)`
-function _is_halfspace(expr::Operation)
+function _is_halfspace(expr::Term)
     got_halfspace = true
 
     # find sense and normalize
-    if expr.op == <
+    if (expr.op == <=) || (expr.op == <)
```
Comment (schillic) [inline src/Sets/HalfSpace.jl:536]:
```suggestion
    if expr.op in (<=, <)
```

Comment (schillic) [inline src/Sets/HalfSpace.jl:536]:
```suggestion
    if expr.op in (<=, <)
```
This is just an aesthetic suggestion. Feel free to ignore it.

File: src/Sets/HalfSpace.jl [hunk 5]
```diff
@@ -529,15 +529,15 @@ return quote
...
         a, b = expr.args
         sexpr = simplify(a - b)
 
-    elseif expr.op == >
+    elseif (expr.op == >=) || (expr.op == >)
```
Comment (schillic) [inline src/Sets/HalfSpace.jl:540]:
```suggestion
    elseif expr.op in (>=, >)
```

**Review by schillic — APPROVED** [review]

File: src/Initialization/init_ModelingToolkit.jl (hunk 2 above)
Comment (schillic) [reply src/Initialization/init_ModelingToolkit.jl:42]: So do you want to change it?

File: src/Initialization/init_ModelingToolkit.jl [hunk 6]
```diff
@@ -1,17 +1,18 @@
...
    """
-       _vec(vars::NTuple{L, Union{<:Operation, <:Vector{Operation}}}) where {L}
+       _vec(vars::NTuple{L, Union{<:Num, <:Vector{Num}}}) where {L}
 
    Transform a tuple of operations into one vector of operations.
 
    ### Input
 
-   - `vars` -- tuple where each element is either an `Operation` or a vector of operations
+   - `vars` -- tuple where each element is either a variable-like (`Num`) or a
```
Comment (mforets) [inline src/Initialization/init_ModelingToolkit.jl:14]:
```suggestion
   - `vars` -- tuple where each element is either variable-like (`Num`) or a
```

File: src/Initialization/init_ModelingToolkit.jl (hunk 2 above)
Comment (YingboMa) [reply src/Initialization/init_ModelingToolkit.jl:42]: You shouldn't dispatch on `Term`. You need to check a expression by testing `istree(expr)`.

**Pull request merged**